### ISSUE TITLE
About training time

### ISSUE THREAD

**Issue by zwyyy215** [jd730/OICR-pytorch#8]
I use a single card GPU, TITANX.training time spent four days, is this normal phenomenon? mAP is 40.2.
What adjustments should I make if I want to speed up my training or improve my mAP?
Thank you for your patient guidance.

**Comment by jd730** [comment]
Since it uses more than 2k rois, we cannot use more batches on a single GPU.
In my case, on TitanXP, I spent a day (45~60 s per 50 iterations).

Although I did not try, I guess that if you change a learning rate as 2e-3 and the [code](https://github.com/jd730/OICR-pytorch/blob/cf203436a163bc96ab8a0a923770d34a8f2ddb80/trainval_net.py#L408) from 2 to 1, it might be the same result boosting the speed.
The current code, following the original code convention, does forward twice and backward once.

For improving performance, you can change the learning rate, weight decay, use different optimizer (I tried Adam and it was bad) and etc....

Or, change the TRAIN.MAX_SIZE which limit the size of images on `lib/model/utils/config.py`

**Comment by jd730** [comment]
The easiest way to improve the performance is let the program run more than 100k. it achieved 43.0 around 100~120k.

**Comment by Mickeybaby** [comment]
Thank you again.I would like to ask if you will follow up other papers on weakly supervised object detection, such as MELM, PCL, CMIL, etc.

**Comment by jd730** [comment]
@Mickeybaby  That's I am not sure, but you can find [PCL](https://github.com/ppengtang/oicr/tree/pcl) and [MELM](https://github.com/vasgaowei/pytorch_MELM)
I have Tang's PCL code modifying to fit my code and it achieves reported performance

**Comment by zwyyy215** [comment]
Thank you！

**Comment by zwyyy215** [comment]
Can you share your pytorch version of PCL code?

**Comment by jd730** [comment]
@zwyyy215 
I am sorry to say that I am not sure whether I can publicize that code since my code is just fitting version of Peng Tang's Pytorch PCL code which means that I am not sure that I can publicize his code until the time that he publicize. However, you can contact him to get the code (PyTorch version).

**Comment by zwyyy215** [comment]
All right, thank you！

**Comment by zwyyy215** [comment]
I am a beginner of weakly supervised object detection and have read some articles recently.Such as :MELM,C-MIL,OICR,PCL and so on.
 I'd like to ask you what else  can do to improve performance. For example, to improve the network structure, to change the loss function, and so on.Can you give me some ideas? Thank you very much.

**Comment by jd730** [comment]
@zwyyy215  There is no definite solution. Maybe you can add a new module, change the network, parameters, and a new loss, etc.

**Comment by zwyyy215** [comment]
Thank you again!

**Comment by zwyyy215** [comment]
Hello, I'm sorry to bother you again. Tang has released the pytorch version of PCL Code.
Unfortunately, his code uses the detectron framework and the data set is in coco format.
I'm a beginner in computer science. I spent most of the month researching his code. I haven't understood it yet.
Tang's pytorch version of PCL has been released. Can your release your code？
I'll be very grateful to you.
Best wishes！

**Comment by ryanLyuJH** [comment]
I used pytorch=0.4.1 which is much faster than pytorch=0.4.0, almost saved 1/2 time. However I don't know why.

**Comment by jd730** [comment]
@ryanLyuJH  Thanks for sharing your experience. Could you tell me your setting more specifically?
When I tested on pytorch==0.40 and 1.0, the speeds are not different.
Thank you.

**Comment by jd730** [comment]
@zwyyy215  I am sorry about replying late. I highly recommend using his code since it is an official pytorch version PCL code. I plan to upload PCL code but I am not sure when I have time to do that.

**Comment by ryanLyuJH** [comment]
Thanks to your good job. I am now studying weakly supervised object detection. I used one GPU, 1080Ti. The setting is just you used. It's about 60-80s for 50 iterations in pytorch0.4.1. However, when I use pytorch0.4.0, it is almost 140s for 50 iterations.

**Comment by bityangke** [comment]
@jd730 Hi have you modified the codebase to pytorch==1.0？
Pytorch==0.4.0 can not run on RTX GPU

**Comment by jd730** [comment]
Hi @bityangke, I also suffered from the same problem as you mentioned. Unfortunately, Cython codes such as NMS, and RoI pooling cannot run on Torch 1.0.

If you need to run on CUDA 10.0 environment, please change them to those in https://github.com/jwyang/faster-rcnn.pytorch/tree/pytorch-1.0 .

Thanks.

**Comment by bityangke** [comment]
for reference, on one single rtx2080ti, when use pytorch 0.4.1， 50 iters cost 50 sec; when use pytorch 1.3, 50 iters cost 28 sec;

**Comment by jd730** [comment]
@bityangke Hi, Thank you for sharing your experience. According to it, the reason may be because of RoI pooling and NMS.
